cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -2183,81 +2183,196 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
+                let is_sat = op == Opcode::FcvtToUintSat || op == Opcode::FcvtToSintSat;
 
-            let src_copy = ctx.alloc_tmp(RegClass::V128, input_ty);
-            ctx.emit(Inst::gen_move(src_copy, src, input_ty));
+                let src_copy = ctx.alloc_tmp(RegClass::V128, input_ty);
+                ctx.emit(Inst::gen_move(src_copy, src, input_ty));
 
-            let tmp_xmm = ctx.alloc_tmp(RegClass::V128, input_ty);
-            let tmp_gpr = ctx.alloc_tmp(RegClass::I64, output_ty);
+                let tmp_xmm = ctx.alloc_tmp(RegClass::V128, input_ty);
+                let tmp_gpr = ctx.alloc_tmp(RegClass::I64, output_ty);
 
-            let srcloc = ctx.srcloc(insn);
-            if to_signed {
-                ctx.emit(Inst::cvt_float_to_sint_seq(
-                    src_size, dst_size, is_sat, src_copy, dst, tmp_gpr, tmp_xmm, srcloc,
-                ));
+                let srcloc = ctx.srcloc(insn);
+                if to_signed {
+                    ctx.emit(Inst::cvt_float_to_sint_seq(
+                        src_size, dst_size, is_sat, src_copy, dst, tmp_gpr, tmp_xmm, srcloc,
+                    ));
+                } else {
+                    ctx.emit(Inst::cvt_float_to_uint_seq(
+                        src_size, dst_size, is_sat, src_copy, dst, tmp_gpr, tmp_xmm, srcloc,
+                    ));
+                }
             } else {
-                ctx.emit(Inst::cvt_float_to_uint_seq(
-                    src_size, dst_size, is_sat, src_copy, dst, tmp_gpr, tmp_xmm, srcloc,
-                ));
+                if op == Opcode::FcvtToSintSat {
+                    // Sets destination to zero if float is NaN
+                    let tmp = ctx.alloc_tmp(RegClass::V128, types::I32X4);
+                    ctx.emit(Inst::xmm_unary_rm_r(
+                        SseOpcode::Movapd,
+                        RegMem::reg(src),
+                        tmp,
+                    ));
+                    ctx.emit(Inst::gen_move(dst, src, input_ty));
+                    let cond = FcmpImm::from(FloatCC::Equal);
+                    ctx.emit(Inst::xmm_rm_r_imm(
+                        SseOpcode::Cmpps,
+                        RegMem::reg(tmp.to_reg()),
+                        tmp,
+                        cond.encode(),
+                        false,
+                    ));
+                    ctx.emit(Inst::xmm_rm_r(
+                        SseOpcode::Pand,
+                        RegMem::reg(tmp.to_reg()),
+                        dst,
+                    ));
+
+                    // Sets top bit of tmp if float is positive
+                    // Setting up to set top bit on negative float values
+                    ctx.emit(Inst::xmm_rm_r(
+                        SseOpcode::Pxor,
+                        RegMem::reg(dst.to_reg()),
+                        tmp,
+                    ));
+
+                    // Convert the float to double.
```

Comment:
```suggestion
                    // Convert the float to double quadword.
```

Reply:
By double I meant double word. This is converting to a double word though not a double quadword. I'll change it to say packed double word.

Reply:
> I think the logic looks right but can we add the CLIF tests that verify these individual instructions? I'm thinking that `simd-conversion-run.clif` and `simd-conversion-legalize.clif` (once converted to being a `test compile` emitting vcode) would be very useful to see that each instruction works correctly and compiles to the sequence we expect.

It definitely passes the SIMD Spectest so I am confident it is correct but let me look to add a file test as well.  👍

Reply:
These instructions are tested in `simd_conversions.wast` but this file has not been enabled in `experimental_x64_should_panic` in the `build.rs` so I don't think any spec tests are running for these instructions. Unfortunately, that spec test also checks `narrow` and `widen` which I found a bit annoying; a lot has to be implemented for the spec test to be enabled. So I guess the CLIF file tests will be the only things checking this until all conversions are implemented.

Reply:
Yes, you're right they aren't enabled by default. I ran them manually though, basically removed all tests except the ones related to the packed float conversion to packed signed int. Separately I've also included the file tests that have tests for this conversion .. commenting out tests packed float to packed unsigned int which isn't supported yet.